pennylane/transforms/diagonalize_measurements.py

```diff
@@ -0,0 +1,345 @@
+
+# pylint: disable=protected-access
+
+_default_supported_obs = frozenset(["PauliZ", "Identity"])
+
+
+def null_postprocessing(results):
+    """A postprocessing function returned by a transform that only converts the batch of results
+    into a result for a single ``QuantumTape``.
+    """
+    return results[0]
+
+
+@transform
+def diagonalize_measurements(tape, supported_base_obs=("PauliZ", "Identity")):
+    """Diagonalize the measurements on a tape if they are not supported. Raises an error if the
+    measurements do not commute.
+
+    Args:
+        tape (QNode or QuantumScript or Callable): The quantum circuit to modify the measurements of.
+        supported_base_obs (Optional, Iterable(Str)): A list of names of supported base observables.
+            Allowed names are 'PauliX', 'PauliY', 'PauliZ' and 'Hadamard'. If no list is provided,
+            the transform will diagonalize everything.
+
+    Returns:
+        qnode (QNode) or tuple[List[QuantumScript], function]: The transformed circuit as described in :func:`qml.transform <pennylane.transform>`.
+
+    .. note::
+        This transform will raise an error if it encounters non-commuting terms. To avoid non-commuting terms in
+        circuit measurements, the :func:`split_non_commuting <pennylane.transforms.split_non_commuting>` transform
+        can be applied.
+
+
+    **Examples:**
+
+    This transform allows us to transform QNode measurements into the measurement basis by adding
+    the relevant diagonalizing gates to the end of the tape operations.
+
+    .. code-block:: python3
+
+        from pennylane.transforms import diagonalize_measurements
+
+        dev = qml.device("default.qubit", wires=2)
+
+        @diagonalize_measurements
+        @qml.qnode(dev)
+        def circuit(x):
+            qml.RY(x[0], wires=0)
+            qml.RX(x[1], wires=1)
+            return qml.expval(qml.X(0) @ qml.Z(1)), qml.var(0.5 * qml.Y(2) + qml.X(0))
+
+    Instead of decorating the QNode, we can also create a new function that yields the same
+    result in the following way:
+
+    .. code-block:: python3
+
+        @qml.qnode(dev)
+        def circuit(x):
+            qml.RY(x[0], wires=0)
+            qml.RX(x[1], wires=1)
+            return qml.expval(qml.X(0) @ qml.Z(1)), qml.var(0.5 * qml.Y(2) + qml.X(0))
+
+        diagonalized_circuit = diagonalize_measurements(circuit)
+
+    Applying the transform appends the relevant gates to the end of the cirucit to allow
+    measurements to be in the Z basis, so the original circuit
+
+    >>> print(qml.draw(circuit)([np.pi/4, np.pi/4]))
+    0: ──RY(0.79)─┤ ╭<X@Z> ╭Var[(0.50*Y)+X]
+    1: ──RX(0.79)─┤ ╰<X@Z> │
+    2: ───────────┤        ╰Var[(0.50*Y)+X]
+
+    becomes
+
+    >>> print(qml.draw(diagonalized_circuit)([np.pi/4, np.pi/4]))
+    0: ──RY(0.79)──H────┤ ╭<Z@Z> ╭Var[(0.50*Z)+Z]
+    1: ──RX(0.79)───────┤ ╰<Z@Z> │
+    2: ──Z─────────S──H─┤        ╰Var[(0.50*Z)+Z]
+
+    >>> diagonalized_circuit([np.pi/4, np.pi/4])
```

Comment:
If we remove lines 61 - 72:
```suggestion
    >>> circuit([np.pi/4, np.pi/4])
```